### PR TITLE
Overlapper

### PR DESCRIPTION
**Pull request by dstreett**
Hey, @joe-angell . Hope all is going well!

So, erm. . . this commit got really out of hand. . . Sorry about that.

The biggest one is looking at the overlapper stuff. Everything else was making stats collecting work and general clean up.

If you wouldn't mind looking at it at some point, no rush though! Thank you! :)

### CODE REVIEW

**Comment by joe-angell** [comment]
hey dude, i'll try to work on this over the next week or so while i'm on vacation, :)

**Comment by dstreett** [comment]
No worries! Take your time, there is no rush. 

Again, I'm sorry for being such a jerk and having a large pr.

File: Super-Deduper/test/testSD.cpp [hunk 1]
```diff
@@ -11,15 +11,15 @@ class SDTest : public ::testing::Test {
 
 TEST_F(SDTest, HashMapLoadTest) {
     
-    std::istringstream in1(readData);
+/*    std::istringstream in1(readData);
```
Comment (joe-angell) [inline Super-Deduper/test/testSD.cpp]: why is this commented?

File: Tab-Converter/src/tab-converter.cpp [hunk 2]
```diff
@@ -29,13 +29,16 @@ typedef std::unordered_map <std::string, size_t> Counter;
...
     while (reader.has_next()) {
-        writer->write(*reader.next());
+        ++c["TotalRecords"];
+        writer_helper(reader.next().get(), pe, se, false, c);
     }
+
 }
 
-template <class T, class Impl>
+/*template <class T, class Impl>
```
Comment (joe-angell) [inline Tab-Converter/src/tab-converter.cpp]: remove commented code

File: common/src/ioHandler.cpp [hunk 3]
```diff
@@ -1,16 +1,88 @@
 #include "ioHandler.h"
 #include <exception>
+#include <unordered_map>
+
+
+void setupCounter(Counter &c) {
```
Comment (joe-angell) [inline common/src/ioHandler.cpp]: not sure if this belongs in ioHandler, might be best to move this to some other common file.

File: common/src/ioHandler.cpp [hunk 4]
```diff
@@ -1,16 +1,88 @@
...
+    c["SE_Right_Trim"] = 0;
+    c["SE_Left_Trim"] = 0;
+    c["Overlap_BPs"] = 0;
+    c["Sins"] = 0;
+    c["Lins"] = 0;
+    c["Replaced"] = 0;
+    c["Ignored"] = 0; 
+}
+
+void write_stats(const std::string &statsFile, const bool &appendStats, const Counter &c, const std::string &program_name) {
```
Comment (joe-angell) [inline common/src/ioHandler.cpp]: same here

File: common/src/ioHandler.h [hunk 5]
```diff
@@ -50,9 +52,9 @@ class HtsOfstream {
     }
     
     HtsOfstream(std::string filename_, bool force_, bool gzip_, bool stdout_) : force(force_), filename(filename_), gzip(gzip_),
-                                                                                std_out(stdout_)  { }
+                                                                                std_out(stdout_)  {out = nullptr; }
     
-    HtsOfstream(std::shared_ptr<std::ostream> out_) : out(out_) { }
+    HtsOfstream(std::shared_ptr<std::ostream> out_) : out(out_) {out = nullptr; }
```
Comment (joe-angell) [inline common/src/ioHandler.h]: here you are overriding out with nullptr?  This should be removed afaik.

File: common/src/ioHandler.h [hunk 6]
```diff
@@ -19,24 +19,26 @@
...
 class HtsOfstream {
 private:
-    bool gzip;
+    std::string filename;
     bool force;
+    bool gzip;
     bool std_out;
-    std::string filename;
-    std::shared_ptr<std::ostream> out = nullptr;
+    std::shared_ptr<std::ostream> out;
```
Comment (joe-angell) [inline common/src/ioHandler.h]: either leave this as is `out = nullptr` or fix this in each constructors initializer list.

File: common/src/ioHandler.h [hunk 7]
```diff
@@ -50,9 +52,9 @@ class HtsOfstream {
     }
     
     HtsOfstream(std::string filename_, bool force_, bool gzip_, bool stdout_) : force(force_), filename(filename_), gzip(gzip_),
-                                                                                std_out(stdout_)  { }
+                                                                                std_out(stdout_)  {out = nullptr; }
```
Comment (joe-angell) [inline common/src/ioHandler.h]: use initializer list

File: common/src/read.h [hunk 8]
```diff
@@ -5,7 +5,10 @@
 #include <boost/optional.hpp>
 #include <memory>
 #include <iostream> 
+#include <unordered_map>
+
 typedef boost::dynamic_bitset<> BitSet;
+typedef std::unordered_map<std::string, size_t> Counter;
```
Comment (joe-angell) [inline common/src/read.h]: you have this typdef defined in multiple headers, best to move this into a common header.

File: Overlapper/src/overlapper.h [hunk 9]
```diff
@@ -0,0 +1,227 @@
...
+}
+
+spReadBase check_read(PairedEndRead &pe , const size_t &maxMis, const size_t &minOver, histVec &insertLength, const bool &stranded, const size_t &checkLengths, const bool &adapterTrimming) {
+    
+    Read &r1 = pe.non_const_read_one();
+    Read &r2 = pe.non_const_read_two();
+
+    bool swapped = false;
+    /*Read1 is always longer than Read 2)*/
+    if (r1.getLength() < r2.getLength()) {
```
Comment (dstreett) [inline Overlapper/src/overlapper.h]: We could, but the only algorithm it directly impacts is the overlapping one, so there isn't really a need.

File: Super-Deduper/test/testSD.cpp (hunk 1 above)
Comment (dstreett) [inline Super-Deduper/test/testSD.cpp]: Because I'm a noob. . . sorry.

File: Overlapper/src/overlapper.h [hunk 10]
```diff
@@ -0,0 +1,227 @@
...
+ * Multi map because a single kemr could appear multiple places*/
+seqLookup readOneMap(const std::string &seq1, size_t step) {
+
+    seqLookup baseReadMap;
+    size_t seqLen = seq1.length();
+
+    for (size_t bp = 0; bp < seqLen - kmer; bp+=kmer) {
+        baseReadMap.insert(std::make_pair(seq1.substr(bp, kmer), bp));
+    }
+    baseReadMap.insert(std::make_pair(seq1.substr(seqLen - kmer, kmer), seqLen - kmer));
```
Comment (dstreett) [inline Overlapper/src/overlapper.h]: seqLen is now equal to `seqLen = seq1.length() - 1` fixing the off by one error.

**Comment by dstreett** [comment]
Hey, @joe-angell, thanks for the comments! I created a utils file for the stats stuff - are you happy with that solution? I figure it can be a catch all for any more common code we might need down the line.

Thanks again for looking over it!

**Comment by joe-angell** [comment]
yeah looks good

**Pull request merged**